editoast/src/modelsv2/projects.rs

```diff
@@ -0,0 +1,257 @@
+use crate::error::Result;
+use crate::modelsv2::Save;
+use crate::views::pagination::{Paginate, PaginatedResponse};
+use crate::DbPool;
+use actix_web::web::Data;
+use async_trait::async_trait;
+use chrono::{NaiveDateTime, Utc};
+use derivative::Derivative;
+use diesel::sql_types::BigInt;
+use diesel::{sql_query, ExpressionMethods, QueryDsl};
+use diesel_async::{AsyncPgConnection as PgConnection, RunQueryDsl};
+use editoast_derive::ModelV2;
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+use crate::models::List;
+
+crate::schemas! {
+    Ordering,
+    ProjectWithStudies,
+    Project,
+    ProjectRow,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Derivative, ModelV2, ToSchema)]
+#[derivative(Default)]
+#[model(table = crate::tables::project)]
+#[model(row(derive(Debug, Clone, Serialize, ToSchema)))]
+pub struct Project {
+    #[schema(value_type = i64)]
+    pub id: i64,
+
+    #[schema(value_type = String)]
+    pub name: String,
+
+    #[derivative(Default(value = "String::new()"))]
+    #[schema(value_type = String)]
+    pub objectives: String,
+
+    #[derivative(Default(value = "String::new()"))]
+    #[schema(value_type = String)]
+    pub description: String,
+
+    #[derivative(Default(value = "String::new()"))]
+    #[schema(value_type = String)]
+    pub funders: String,
+
+    #[derivative(Default(value = "0"))]
+    #[schema(value_type = i32)]
+    pub budget: i32,
+
+    #[schema(value_type = NaiveDateTime)]
+    pub creation_date: NaiveDateTime,
+
+    #[derivative(Default(value = "Utc::now().naive_utc()"))]
+    pub last_modification: NaiveDateTime,
+
+    #[derivative(Default(value = "Vec::new()"))]
+    #[schema(value_type = Vec<String>)]
+    pub tags: Vec<Option<String>>,
```

Comment:
It feels like it should be `Vec<String>`.

```suggestion
    pub tags: Vec<String>,
```

Reply:
Done.

Reply:
Ideally, we should migrate this type to `Array<Text>`

Reply:
PostgreSQL does not support to create an array of non-null text values (`Array<Text>`).
the following script does not work:
```
CREATE TYPE non_null_text AS (
    text_value TEXT NOT NULL
);

CREATE TABLE test (
    id SERIAL PRIMARY KEY,
    tags non_null_text[] NOT NULL
);
```
postgres supports only `TEXT[] NOT NULL`

Reply:
PostgreSQL doesn't allow types like `(TEXT NOT NULL)[]` so arrays are always nullable by default. We can prevent this by adding a `CHECK` constraint, but that will be missed by diesel-cli. I suggest using `#[model(remote = TextArray)]` and keep this field a `Vec<String>`. There's probably some tweaking to do on `TextArray` to achieve that.

Reply:
I have created a new struct `Tags` instead of `TextArray`. Thanks @leovalais